Replace `snipes` with `cleanByImpersonation` function

espendk opened the pull request:
This PR removes `snipes` and `snipesFor` from Mangrove and replaces them with a `cleanByImpersonation` function that only allows failing offers to be executed.

# Major changes

In addition to the primary change (not allowing snipes in general) the following changes have been made:

## `MgvCleaner` has been removed
Mangrove core now provides the cleaning functionality that `MgvCleaner` used to provide, so it `MgvCleaner` has been removed.

##  Cleaning of multiple offers is more resilient
Before, `MgvCleaner.collect*` would revert if execution of any offer succeeded. The new `cleanByImpersonation` is more resilient as it will not revert the whole tx if an offer succeeds but will revert just the execution of that offer and continue cleaning of the next offers. 

Thus a cleaner that attempts to clean multiple offers in one tx will be able to successfully clean some offers despite some offers failing to be cleaned.

## Taker impersonation no longer checks permits
The new `cleanByImpersonation` function has integrated the `taker` parameter of the old `snipesFor`. This allows cleaning by impersonation, ie where another taker's funds and approval of Mangrove for transfers of quote can be used to execute an offer and make it fail (which was already possible via `MgvCleaner.collectByImpersonation`).

This has been integrated into the `cleanByImpersonation` function instead of having a separate `cleanFor` function because
1. no permit/allowance is needed nor does it have to be checked and
2. cleaners are expected to be advanced users, so requiring an explicit `taker` parameter with non-obvious semantics seems ok.

# Migration of tests away from snipes
Many tests rely on sniping. To allow gradual migration away from this, the snipe functions have been moved to a `TestSnipingMangrove` contract that `MangroveTest` provides to tests.

To make clear that these tests do something non-standard, the snipe functions have been renamed to `snipesInTest` and `snipesForInTest`.

Migration of tests will be done in later PR's as merging this PR will (1) hopefully stop further use of snipes in tests and (2) enable that we can start updating downstream components to the new MGV V2 API.

espendk committed:
fix(doc): Small correction to documentation in `execute`

espendk committed:
feat(offerTaking): Add clean function to Mangrove

espendk committed:
feat(MgvCleaner): Revert only succeeding offer executions instead of whole tx

espendk committed:
feat(offerTaking): Remove `snipes` and `snipesFor`

Many tests rely on sniping. To allow gradual migration away from this, the snipe functions have been moved to a `TestSnipingMangrove` contract that `MangroveTest` provides to tests.
To make clear that these tests do something non-standard, the snipe functions have been renamed to `snipesInTest` and `snipesForInTest`.

adhusson commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
-      feePaid = mor.feePaid;
-    }
-  }
+      /* If we removed the `isLive` conditional, a single expired or nonexistent offer in `targets` would revert the entire transaction (by the division by `offer.gives` below since `offer.gives` would be 0). We also check that `gasreq` is not worse than specified. A taker who does not care about `gasreq` can specify any amount larger than $2^{24}-1$. A mismatched price will be detected by `execu ...
```
```suggestion
      /* FIXME: edit comment: If we removed the `isLive` conditional, a single expired or nonexistent offer in `targets` would revert the entire transaction (by the division by `offer.gives` below since `offer.gives` would be 0). We also check that `gasreq` is not worse than specified. A taker who does not care about `gasreq` can specify any amount larger than $2^{24}-1$. A mismatched price will be detected by `execute`. */
```

adhusson commented on src/MgvOfferTakingWithPermit.sol:
```diff
@@ -124,23 +124,17 @@ abstract contract MgvOfferTakingWithPermit is MgvOfferTaking {
-      deductSenderAllowance(outbound_tkn, inbound_tkn, taker, takerGave);
-    }
-  }
+  // FIXME: This doesn't use permit or allowance at all, so perhaps it's better to move it to MgvOfferTaking? Or simply expose the `taker` parameter there?
```
Removing `deductSenderAllowance` is scary even though I do not see an attack right now.

adhusson replied:
I would rather keep it in :)

adhusson replied:
@espendk convinced me

espendk committed:
docs: Add FIXME to update a comment

Co-authored-by: Adrien Husson <adhusson@gmail.com>

espendk commented:
~~Just realized that the old snipe tests that test the details of execution (eg events emitted and transfers made) should be migrated as part of this PR. Will look at that Monday!~~

On second thought (and seeing #444 building on this PR) I think we should merge this after review and include the offer execution tests in the upcoming test migration PR. The feature will still be appropriately tested before merging to `develop`.

lnist commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
+  // FIXME: Document cleaning
+  // FIXME: Maybe `cleanFailingOffer` instead?
+  /* Any `taker` can be impersonated when cleaning because the function reverts if the offer succeeds, cancelling any token transfers. And after a `clean` where the offer has failed, all token transfers have been reverted -- but the sender will still have received the bounty of the failing offers. */
+  // FIXME: This would be simpler if tick had to match exactly: Then `fillWants` wouldn't be needed and `fillVolume` could be simplified to either `takerWants` or `takerGives`.
```
Do we have a good use case for non-exact match? Simplified code seems like a win.

adhusson replied:
I'm wondering about the opposite: should we completely remove the price match? Initially it's there for "normal snipes" (to protect the taker). In the absence of normal snipes, what happens if we just ignore the price?

espendk replied:
I think the cleaner risks spending more gas on failing to make the offer fail: The offer is now more expensive and no longer fails.

lnist replied:
The scenario is that the maker updates the offer between us making the decision to clean and actually cleaning - I don't think we can a priori say anything about whether it would still fail. Skipping early saves gas - in case of failure still, we win gas. I don't know what is most common.
I lean towards skipping since information has changed - spending gas on executing the entire offer if it succeeds is quite wasteful.

espendk replied:
I went for the strict match and `takerWants` instead of `fillVolume` and `fillWants` in [92eff73](https://github.com/mangrovedao/mangrove-core/pull/437/commits/92eff7372aa66d7dd695a6b7a741334d3f9add01)

lnist commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
-            // updates best&tick
-            sor.local = dislodgeOffer(pair, sor.offer, sor.local, true);
-          }
+        /* In the market order, we were able to avoid stitching back offers after every `execute` since we knew a continuous segment starting at best would be consumed. Here, we cannot do this optimisation since offers in the `targets` array may be anywhere in the book. So we stitch together offers immediately after each `execute`. */
```
also a comment that needs to be updated. You added a FIXME further up, maybe also here? Though I suspect the code has quite a bit of outdated comments now.

espendk replied:
All of the comments in that file should be updated :-) I made a general comment on this (and @adhusson added a specific one) and I think it'll be easiest to do it in one go once the `feat/ticks` has stabilized. But I can be persuaded otherwise :-)

lnist commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
-            postExecute(mor, sor, gasused, makerData, mgvData);
-          }
+        /* In an inverted Mangrove, amounts have been lent by each offer's maker to the taker. We now call the taker. This is a noop in a normal Mangrove. */
+        // FIXME: Is this needed when offer execution failed? I don't think so, but should be covered by tests
```
Do we leave this FIXME behind?

espendk replied:
Good point, I'll remove it.

adhusson replied:
To answer the FIXME: `executeEnd` is a noop on Mangrove anyway but is useful on InvertedMangrove (it calls the taker contract).

espendk replied:
So `executeEnd` should still be invoked when an offer fails, right?

espendk replied:
Reintroduce `executeEnd` in [76393f2](https://github.com/mangrovedao/mangrove-core/pull/437/commits/76393f2b9000721d1456705f839d761e0d1e9244)

lnist commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
+      bounty = mor.totalPenalty;
+
+      /* Over the course of the snipes order, a penalty reserved for `msg.sender` has accumulated in `mor.totalPenalty`. No actual transfers have occured yet -- all the ethers given by the makers as provision are owned by Mangrove. `sendPenalty` finally gives the accumulated penalty to `msg.sender`. */
+      sendPenalty(bounty);
```
Previously we could clean multiple offers and then in the end do the transfer (at cast 21000 gas). We now pay that amount of gas for each transfer, thus upping the needed penalty to cover the cost.

Should gasbase compensate a normal market order taker, or a cleaner? If it is to incentivize cleaning, then that is the important one.

adhusson replied:
I agree -- gasbase should compensate max(market order overhead, cleaning overhead). 

I didn't get the 21k gas reference?

espendk replied:
Good point!

I think this is a strong reason for moving the multi-clean functionality from `MgvCleaner` to the core. Should be pretty quick to do.

Unless @adhusson objects, I'll go ahead and make the change.

espendk replied:
Done in [e4cfb66](https://github.com/mangrovedao/mangrove-core/pull/437/commits/e4cfb668bd6ebfb2547952379be91c8c510a1482)

lnist commented on src/MgvOfferTaking.sol:
```diff
@@ -306,146 +306,97 @@ abstract contract MgvOfferTaking is MgvHasOffers {
-          if (mgvData == "mgv/tradeSuccess") {
-            successCount += 1;
-          }
+        require(mgvData != "mgv/tradeSuccess", "mgv/clean/offerDidNotFail");
```
This looks correct - however, would it be safer to explicitly check the allowed failure values? 
`require(mgvData == "mgv/makerRevert" || mgvData == "mgv/makerTransferFail" || mgvData == "mgv/makerReceiveFail", "mgv/clean/offerDidNotFail");

adhusson replied:
The intent was to be passively future-proof against additional errors that would be added later. I think that still applies so I would tend towards keeping it as is.

espendk replied:
Hmm, the point here is to revert explicitly if the offer succeeds, because then it shouldn't be cleaned.
But maybe that isn't clear?

lnist commented on src/periphery/MgvCleaner.sol:
```diff
@@ -3,17 +3,15 @@ pragma solidity ^0.8.10;
    2) Run `collect` on the offers that you detected were failing.
 
-   You can adjust takerWants/takerGives and gasreq as needed.
+   Instead of using your own address, you can use `collectByImpersonation` to impersonate another taker (`takerToImpersonate`) who has approved Mangrove for `inbound_tkn`. This allows borrowing that taker's `inbound_tkn` funds for cleaning instead of using `msg.sender`'s funds (who need not have any).
```
takerToImpersonate is now just `taker`, right?

espendk replied:
Indeed, I'll fix.

lnist commented on test/core/Gatekeeping.t.sol:
```diff
@@ -607,13 +607,52 @@ contract GatekeepingTest is IMaker, MangroveTest {
+
   /* Snipe failure */
 
   function snipesKO(uint id) external {
```
is snipesKO replaces by cleanKO?

espendk replied:
Yes, I missed a rename when adapting the reentrancy tests and fixed on the test migration branch - I didn't expect you to notice ;-)

I'll port the fix here!

adhusson reviewed (approved):
lgtm pending comments above

espendk committed:
chore: Address review comments

espendk committed:
chore: Address review comment

espendk committed:
fix: Migrate cleaning gas test to new API

espendk committed:
chore: Silence irrelevant warning about unused variable in override

espendk committed:
feat: Move multi-offer cleaning from MgvCleaner to Mangrove

espendk committed:
chore: Address review comment

espendk committed:
refactor: Rename `clean` to `cleanByImpersonation`

espendk committed:
feat: Require precise tick and takerWants for cleaning

Cleaning will now fail early if the tick does not match exactly to protect the cleaner from trying to clean an offer that has been updated.
Also, cleaning must now always specify takerWants; With the tick being fixed, the ability to specify wants or gives does not add expressivity, so the code and gas can be saved.

espendk committed:
Merge branch 'feat/ticks' into feat/clean-instead-of-snipe

espendk committed:
fix: Missed rename in 84ffe72

espendk committed:
fix: Migrate monitor scripts to clean instead of snipe

Required after merge from `develop`.
